linera-core/src/worker.rs

```diff
@@ -726,14 +766,64 @@ where
             notify_when_messages_are_delivered,
         )
         .await;
-        self.cache_recent_value(Cow::Owned(certificate.value)).await;
+        self.cache_recent_hashed_certificate_value(Cow::Owned(certificate.value))
+            .await;
 
         #[cfg(with_metrics)]
         NUM_BLOCKS.with_label_values(&[]).inc();
 
         Ok((info, actions))
     }
 
+    /// Returns an error if the block requires a blob we don't have, or if unrelated blobs were provided.
+    async fn check_no_missing_blob(
+        &self,
+        mut required_blob_ids: HashSet<BlobId>,
+        hashed_blobs: &[HashedBlob],
+        pending_blobs: &BTreeMap<BlobId, HashedBlob>,
+    ) -> Result<(), WorkerError> {
+        // Find all certificates containing blobs used when executing this block.
+        for hashed_blob in hashed_blobs {
+            let blob_id = hashed_blob.id();
+            ensure!(
+                required_blob_ids.remove(&blob_id),
+                WorkerError::UnneededBlob { blob_id }
+            );
+        }
+
+        let recent_blobs = self.recent_hashed_blobs.lock().await;
+        let missing_blob_ids = required_blob_ids
+            .into_iter()
+            .filter(|blob_id| {
+                !recent_blobs.contains(blob_id) && !pending_blobs.contains_key(blob_id)
```

Comment:
(This can probably get the same optimization as in https://github.com/linera-io/linera-protocol/pull/2038?)